os/mm/mm_heap/mm_mallinfo.c

```diff
@@ -111,9 +113,33 @@ int mm_mallinfo(FAR struct mm_heap_s *heap, FAR struct mallinfo *info)
 
 		mm_takesemaphore(heap);
 
-		for (node = heap->mm_heapstart[region]; node < heap->mm_heapend[region]; node = (struct mm_allocnode_s *)((char *)node + node->size)) {
+		prev = NULL;
+		node = heap->mm_heapstart[region];
+		next = (struct mm_allocnode_s *)((char *)node + node->size);
+
+		for (; node < heap->mm_heapend[region]; prev = node, node = next, next = (struct mm_allocnode_s *)((char *)next + next->size)) {
 			mvdbg("region=%d node=%p size=%p preceding=%p (%c)\n", region, node, node->size, (node->preceding & ~MM_ALLOC_BIT), (node->preceding & MM_ALLOC_BIT) ? 'A' : 'F');
 
+			if ((prev && prev->size != node->preceding & ~MM_ALLOC_BIT) ||
+				(node->size != next->preceding & ~MM_ALLOC_BIT) ||
+				(!(node->preceding & ~MM_ALLOC_BIT) &&
+				(((struct mm_freenode_s*)node)->blink->flink != node || ((struct mm_freenode_s *)node)->flink->blink != node))
```

Comment:
Sorry, need to check null for `node->blink` and `node->flink` for start or end of free node llink